### PR TITLE
add: Security context constraints

### PR DESCRIPTION
**Pull request by VanillaSpoon**
## Why are these changes needed?

This pr adds the Kuberay SCC to the repository to be deployed. The changes are added to both the Kustomize deployments.

## Related issue number

closes https://github.com/red-hat-data-services/distributed-workloads/issues/15

## To check changes

Update the ray-operator/Makefile deploy to :
```
deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
	cd config/manager && $(KUSTOMIZE) edit set image kuberay/operator=${IMG}
	($(KUSTOMIZE) build config/openshift | kubectl create -f -) || ($(KUSTOMIZE) build config/openshift | kubectl replace -f -)
```

Kustomize:
  - From within the ray-operator directory, run `make deploy -e IMG=<your-image`
  
  Check to ensure the run-as-ray-user SCC has been applied correctly, and ensure Users is set in the correct namespace (opendatahub):
 ```
  Users:                                                                    │
   system:serviceaccount:opendatahub:kuberay-operator  
 ```
 
 Also ensure the ray-operator, and rbacs are applied within the `opendatahub` namespace.
 
## Checks

- [x]  I've made sure the tests are passing.
- Testing Strategy
    - [ ]  Unit tests
    - [x]  Manual tests
    - [ ]  This PR is not tested :(

### CODE REVIEW

**Comment by anishasthana** [comment]
Since we are only using the kustomize manifests for the ODH operator, I don't think we should change the existing helm manifests, just to make future rebases easier. @astefanutti wdyt?

File: ray-operator/config/default/kustomization.yaml [hunk 1]
```diff
@@ -12,16 +12,17 @@ namespace: ray-system
 #commonLabels:
 #  someName: someValue
 
-bases:
```
Comment (astefanutti) [inline ray-operator/config/default/kustomization.yaml:15]: Let's not commit that change as it's not strictly related to the SCC patch.

File: ray-operator/config/rbac/kustomization.yaml [hunk 2]
```diff
@@ -5,6 +5,10 @@ resources:
 - leader_election_role_binding.yaml
 - service_account.yaml
 
-commonLabels:
-  app.kubernetes.io/name: kuberay
-  app.kubernetes.io/component: kuberay-operator
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (astefanutti) [inline ray-operator/config/rbac/kustomization.yaml:8]: Is that change really needed?

File: helm-chart/kuberay-operator/values.yaml [hunk 3]
```diff
@@ -17,6 +17,9 @@ serviceAccount:
   # If not set and create is true, a name is generated using the fullname template
   name: "kuberay-operator"
 
+scc:
```
Comment (astefanutti) [inline helm-chart/kuberay-operator/values.yaml:20]: I'd removed the Helm related changes as suggested by @anishasthana

File: ray-operator/config/openshift/ray_operator_scc.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+apiVersion: security.openshift.io/v1
+metadata:
+  name: run-as-ray-user
+seLinuxContext:
+  type: MustRunAs
+runAsUser:
+  type: MustRunAs
+  uid: 1000
+users:
+  - 'system:serviceaccount:$(namespace):kuberay-operator'
```
Comment (astefanutti) [inline ray-operator/config/openshift/ray_operator_scc.yaml:11]: Where is the namespace variable set?

**Review by astefanutti — CHANGES_REQUESTED** [review]

File: ray-operator/config/openshift/ray_operator_scc.yaml (hunk 4 above)
Comment (VanillaSpoon) [reply ray-operator/config/openshift/ray_operator_scc.yaml:11]: Good spot, thanks @astefanutti 
The namespace is set within the default kustomize file, however, this won't translate to the openshift kustomize.

File: helm-chart/kuberay-operator/values.yaml (hunk 3 above)
Comment (VanillaSpoon) [reply helm-chart/kuberay-operator/values.yaml:20]: Yea, these were added for consistency as I believe the CI tests make use of helm for deployment? However, as these tests also run on kind there is no need for the SCC anyway.

File: ray-operator/config/rbac/kustomization.yaml (hunk 2 above)
Comment (VanillaSpoon) [reply ray-operator/config/rbac/kustomization.yaml:8]: These were changes made due to a Warning I received during the kustomize deployments regarding deprecations, I followed the instruction to clean up any other discrepancies. However this could be addressed separately:
```
# Warning: 'bases' is deprecated. Please use 'resources' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
```

Comment (astefanutti) [reply ray-operator/config/rbac/kustomization.yaml:8]: Right, ideally these changes should be contributed to KubeRay upstream, and rebase our fork on top. We want to keep the patches as small as possible in the later.

Comment (VanillaSpoon) [reply ray-operator/config/rbac/kustomization.yaml:8]: Perfect, thanks @astefanutti :)

File: ray-operator/config/openshift/params.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+varReference:
+  - path: subjects[]/namespace
+    kind: ClusterRoleBinding
```
Comment (astefanutti) [inline ray-operator/config/openshift/params.yaml:3]: I don't think it's need anymore as my understanding is it was required for the binding to the edit ClusterRole which you removed.

File: ray-operator/config/openshift/params.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+  - path: users[]
+    kind: SecurityContextConstraints
+  - path: spec/template/spec/containers[]/image
```
Comment (astefanutti) [inline ray-operator/config/openshift/params.yaml:6]: I think the corresponding variable for parameterising the image is missing in `param.env`.

File: ray-operator/config/openshift/params.yaml (hunk 5 above)
Comment (VanillaSpoon) [reply ray-operator/config/openshift/params.yaml]: Thanks @astefanutti :)

File: ray-operator/config/openshift/params.yaml (hunk 6 above)
Comment (VanillaSpoon) [reply ray-operator/config/openshift/params.yaml:4]: Thanks @astefanutti  :)
I have address these changes in bda215576e97c9488c56e50bc1ae99fc61b2eddd

**Comment by astefanutti** [comment]
opendatahub-io/opendatahub-operator#619 has been merged so the required namespace variable is injected by the ODH operator.

File: ray-operator/config/openshift/params.yaml (hunk 6 above)
Comment (astefanutti) [reply ray-operator/config/openshift/params.yaml:4]: I think we would need to have that variable actually used. We used to duplicate the entire manifests in the DW repository: https://github.com/opendatahub-io/distributed-workloads/tree/2049ceba5bd72efd6c49a189c6d0af2f0eca8a98/ray/operator

But now, I think we could only add a patch that overrides the KubeRay Deployment image in this overlay. We just need to make sure variables are correctly taken into account in patches.

Comment (VanillaSpoon) [reply ray-operator/config/openshift/params.yaml:4]: Hey @astefanutti 
Thank you for this. 

Just to clarify, would I be safe to remove params.yaml and params.env, as well as: 
```
configMapGenerator:
- name: ray-config
  envs:
    - params.env
configurations:
  - params.yaml

vars:
- name: namespace
  objref:
    kind: ConfigMap
    name: ray-config
    apiVersion: v1
  fieldref:
    fieldpath: data.namespace
```
Allowing us to set the image in kusomization.yaml and have the namspace injected by the ODH operator. Leaving us with:
    
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: opendatahub

resources:
- ray_operator_scc.yaml
- ../default

images:
- name: kuberay/operator
  newName: quay.io/kuberay/operator
  newTag: v0.6.0
```

```
kind: SecurityContextConstraints
apiVersion: security.openshift.io/v1
metadata:
  name: run-as-ray-user
seLinuxContext:
  type: MustRunAs
runAsUser:
  type: MustRunAs
  uid: 1000
users:
  - 'system:serviceaccount:$(namespace):kuberay-operator'

```

Comment (astefanutti) [reply ray-operator/config/openshift/params.yaml:4]: @VanillaSpoon I don't think the namespace plugin will interpolate the `$(namespace)` variable in the SCC. But that has to be checked / tested.

For the image, my understand is that the ODH operator passes custom images as variables, not by setting `images` in the `kustomization.yaml` file.

I can see how your suggestion could simplify things, but I fear it won't be possible. Maybe we can keep the parameters approach in the short term, and reconsider it once we have that first working version?

Comment (VanillaSpoon) [reply ray-operator/config/openshift/params.yaml:4]: Thanks again @astefanutti 
I've pushed some updates to f9cbdefc4832dc3c3779dc53ad7282e43fb77f96 :)

**Comment by astefanutti** [comment]
Thanks for the thorough testing @VanillaSpoon.
